## Let's get some GPT in here. Dale 

![gpt](../Images/chatgpt.jpg)

Some installs/upgrades to make sure everything was up to date for the GPT code to work properly: 

In [1]:
#!pip install --upgrade tensorflow
#!pip install --upgrade transformers
#!pip install --upgrade tokenizers

In [2]:
#https://huggingface.co/docs/transformers/installation 
#!pip install 'transformers[tf-cpu]'

In [3]:
#!pip install accelerate -U

### Causal Language Modeling with DistilGPT2 

Instructions [here](https://huggingface.co/docs/transformers/v4.22.2/en/tasks/language_modeling) and [here](https://github.com/christianversloot/machine-learning-articles/blob/main/easy-causal-language-modeling-with-machine-learning-and-huggingface-transformers.md)

In [4]:
from transformers import GPT2Tokenizer, GPT2LMHeadModel, TextDataset, DataCollatorForLanguageModeling, Trainer, TrainingArguments

# Load tokenizer
tokenizer = GPT2Tokenizer.from_pretrained('distilgpt2')

# Tokenize the training data
train_dataset = TextDataset(
    tokenizer=tokenizer,
    file_path='../Data/pitbull.txt',
    block_size=128
)

# Define the data collator for language modeling
data_collator = DataCollatorForLanguageModeling(
    tokenizer=tokenizer,
    mlm=False
)

# Load the GPT-2 model
model = GPT2LMHeadModel.from_pretrained('distilgpt2')

# Define the training arguments
training_args = TrainingArguments(
    output_dir='../GPT/',  
    overwrite_output_dir=True,
    num_train_epochs=3,
    per_device_train_batch_size=16,
    save_steps=1000,
    save_total_limit=2,
    prediction_loss_only=True,
    logging_dir='../GPT/' 
)

# Create the Trainer
trainer = Trainer(
    model=model,
    args=training_args,
    data_collator=data_collator,
    train_dataset=train_dataset,
)

# Start the training
trainer.train()


2023-06-10 03:42:08.441133: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
/Users/danner/opt/anaconda3/lib/python3.9/site-packages/transformers/data/datasets/language_modeling.py:53: FutureWarning: This dataset will be removed from the library soon, preprocessing should be handled with the 🤗 Datasets library. You can have a look at this example script for pointers: https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_mlm.py
  warnings.warn(
/Users/danner/opt/anaconda3/lib/python3.9/site-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=

TrainOutput(global_step=636, training_loss=3.8688333859233737, metrics={'train_runtime': 13944.4688, 'train_samples_per_second': 0.728, 'train_steps_per_second': 0.046, 'total_flos': 331389603938304.0, 'train_loss': 3.8688333859233737, 'epoch': 3.0})

 First Time
 
[636/636 3:38:30, Epoch 3/3]
Step	Training Loss
500	3.922000
TrainOutput(global_step=636, training_loss=3.8688333859233737, metrics={'train_runtime': 13140.3418, 'train_samples_per_second': 0.772, 'train_steps_per_second': 0.048, 'total_flos': 331389603938304.0, 'train_loss': 3.8688333859233737, 'epoch': 3.0})

Second Time

 [636/636 3:01:21, Epoch 3/3]
Step	Training Loss
500	3.922000
TrainOutput(global_step=636, training_loss=3.8688333859233737, metrics={'train_runtime': 10900.8354, 'train_samples_per_second': 0.931, 'train_steps_per_second': 0.058, 'total_flos': 331389603938304.0, 'train_loss': 3.8688333859233737, 'epoch': 3.0})

Third Time

[636/636 3:52:04, Epoch 3/3]
Step	Training Loss
500	3.922000
TrainOutput(global_step=636, training_loss=3.8688333859233737, metrics={'train_runtime': 13944.4688, 'train_samples_per_second': 0.728, 'train_steps_per_second': 0.046, 'total_flos': 331389603938304.0, 'train_loss': 3.8688333859233737, 'epoch': 3.0})

Over 3 1/2 hours 😍 

In [5]:
# Save the model that took way too long 
model.save_pretrained('../GPT')

#https://stackoverflow.com/questions/66276186/huggingface-gpt2-tokenizer-configuration-in-config-json 
tokenizer.save_pretrained('../GPT')

('../GPT/tokenizer_config.json',
 '../GPT/special_tokens_map.json',
 '../GPT/vocab.json',
 '../GPT/merges.txt',
 '../GPT/added_tokens.json')

![pc](../Images/kickpc.WEBP)